src/pip/_internal/wheel.py

```diff
@@ -391,9 +393,20 @@ def _get_script_text(entry):
 
     # Generate the console and GUI entry points specified in the wheel
     if len(console) > 0:
-        generated.extend(
-            maker.make_multiple(['%s = %s' % kv for kv in console.items()])
+        generated_console_scripts = maker.make_multiple(
+            ['%s = %s' % kv for kv in console.items()]
         )
+        generated.extend(generated_console_scripts)
+        for destfile in generated_console_scripts:
+            parent_dir = os.path.dirname(destfile)
+            if parent_dir not in _ENV_PATH_PARTS and warn_script_location:
+                msg = (
+                    "Script '%s' is not installed in a directory on PATH. "
+                    "Consider adding %s to PATH. "
+                    "Use --no-warn-script-location to suppress this message."
+                )
+                logger.warn(msg, os.path.basename(destfile), parent_dir)
```

Comment:
Instead of having a warning for each script that's installed, would it be better to group them by parent_dir? So we have

> Scripts A, B and C were installed in FOO, which is not on PATH.
> Scripts D, E and F were installed in BAR, which is not on PATH.
> Consider adding these directories to PATH, or if you prefer to suppress this warning, use --no-warn-script-location.

The code will be a bit more complex, but I think the end result for the user will be better.

Otherwise, this LGTM.